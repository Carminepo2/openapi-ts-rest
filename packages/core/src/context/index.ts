import { type OpenAPIObject } from "openapi3-ts/oas30";

import { makeRefObjectResolvers } from "./makeRefObjectResolvers";
import { processObjectSchemas } from "./processObjectSchemas";
import { processOperationsObjects } from "./processOperationsObjects";

export interface Context {
  openAPIDoc: OpenAPIObject;
  operationObjects: ReturnType<typeof processOperationsObjects>;
  resolveObject: ReturnType<typeof makeRefObjectResolvers>["resolveObject"];
  resolveRef: ReturnType<typeof makeRefObjectResolvers>["resolveRef"];
  schemasToExportMap: ReturnType<typeof processObjectSchemas>["schemasToExportMap"];
  topologicallySortedSchemas: ReturnType<typeof processObjectSchemas>["topologicallySortedSchemas"];
}

export function generateContext(openAPIDoc: OpenAPIObject): Context {
  const { resolveObject, resolveRef } = makeRefObjectResolvers(openAPIDoc);
  const operationObjects = processOperationsObjects(openAPIDoc, resolveObject);
  const { schemasToExportMap, topologicallySortedSchemas } = processObjectSchemas(
    openAPIDoc,
    resolveRef
  );

  return {
    openAPIDoc,
    operationObjects,
    resolveObject,
    resolveRef,
    schemasToExportMap,
    topologicallySortedSchemas,
  };
}
