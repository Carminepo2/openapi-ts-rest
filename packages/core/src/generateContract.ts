import type { OpenAPIObject } from "openapi3-ts";
import type { Options as PrettierOptions } from "prettier";

import SwaggerParser from "@apidevtools/swagger-parser";

import { createContext } from "./context/createContext.js";
import { apiOperationToAstTsRestContract } from "./converters/apiOperationToAstTsRestContract.js";
import { schemaObjectToAstZodSchema } from "./converters/schemaObjectToAstZodSchema.js";
import { getApiOperationObjects } from "./getApiOperationObjects.js";
import { getExportedSchemas } from "./getTopologicallySortedSchemas.js";
import { prettify } from "./lib/prettier.js";
import {
  tsChainedMethodCall,
  tsFunctionCall,
  tsNamedImport,
  tsNewLine,
  tsObject,
  tsVariableDeclaration,
} from "./lib/ts.js";
import { AstTsWriter } from "./lib/utils.js";

export interface GenerateContractOptions {
  /**
   * The OpenAPI schema to generate the ts-rest contract from.
   * It can be either an OpenAPI schema object or a URL to an OpenAPI schema.
   */
  openApi: OpenAPIObject | string;
  /**
   * The Prettier configuration to use for formatting the generated code.
   */
  prettierConfig?: PrettierOptions | null;
}

/**
 * Generates a ts-rest contract from an OpenAPI schema.
 * @param {generateContractOptions} options - The options for the generation.
 */
export async function generateContract({
  openApi,
  prettierConfig,
}: GenerateContractOptions): Promise<string> {
  const openApiSchema = (await SwaggerParser.bundle(openApi as never)) as OpenAPIObject;

  const ctx = createContext(openApiSchema);

  const ast = new AstTsWriter();

  ast
    // import { initContract } from "@ts-rest/core";
    .add(tsNamedImport({ from: "@ts-rest/core", import_: ["initContract"] }))
    // import { z } from "zod";
    .add(tsNamedImport({ from: "zod", import_: ["z"] }))
    .add(tsNewLine())
    // const c = initContract();
    .add(tsVariableDeclaration("const", "c", { eq: tsFunctionCall("initContract") }))
    .add(tsNewLine());

  const schemasToExport = getExportedSchemas(ctx);

  if (schemasToExport.length > 0) {
    // Generates the Zod schemas for each component schema.
    for (const { normalizedIdentifier, schema } of schemasToExport) {
      // const [identifier] = z.object({ ... }) | z.string() | z.number() | ...
      ast.add(
        tsVariableDeclaration("const", normalizedIdentifier, {
          eq: schemaObjectToAstZodSchema(schema, ctx),
        })
      );
    }

    ast.add(tsNewLine());

    // export const schemas = { schema1, schema2, ... };
    ast
      .add(
        tsVariableDeclaration("const", "schemas", {
          eq: tsObject(
            ...schemasToExport.map(
              ({ normalizedIdentifier }) => [normalizedIdentifier] satisfies [string]
            )
          ),
          export_: true,
        })
      )
      .add(tsNewLine());
  }

  const apiOperationObjects = getApiOperationObjects(ctx);

  const tsRestAstContracts = apiOperationObjects.map((operationObject) =>
    apiOperationToAstTsRestContract(operationObject, ctx)
  );

  // export const contract = c.router({ ... });
  ast.add(
    tsVariableDeclaration("const", "contract", {
      eq: tsChainedMethodCall("c", ["router", tsObject(...tsRestAstContracts)]),
      export_: true,
    })
  );

  return await prettify(ast.toString(), prettierConfig);
}
