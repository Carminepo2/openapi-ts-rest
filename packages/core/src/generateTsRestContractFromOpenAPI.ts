import type { OpenAPIObject } from "openapi3-ts/oas30";
import type { Options as PrettierOptions } from "prettier";

import SwaggerParser from "@apidevtools/swagger-parser";

import { generateContext } from "./context/index.js";
import { apiOperationToAstTsRestContract } from "./converters/apiOperationToAstTsRestContract.js";
import { schemaObjectToAstZodSchema } from "./converters/schemaObjectToAstZodSchema.js";
import { getApiOperationObjects } from "./getApiOperationObjects.js";
import { prettify } from "./lib/prettier.js";
import {
  tsChainedMethodCall,
  tsFunctionCall,
  tsNamedImport,
  tsNewLine,
  tsObject,
  tsVariableDeclaration,
} from "./lib/ts.js";
import { AstTsWriter } from "./lib/utils.js";
import { processSchemasForExport } from "./processSchemasForExport.js";

interface GenerateTsRestContractFromOpenAPIOptions {
  input: string;
  prettierConfig?: PrettierOptions | null;
}

/**
 * Generates a ts-rest contract from an OpenAPI schema.
 * @param {GenerateTsRestContractFromOpenAPIOptions} options - The options for the generation.
 */
export async function generateTsRestContractFromOpenAPI({
  input,
  prettierConfig,
}: GenerateTsRestContractFromOpenAPIOptions): Promise<string> {
  const openApiSchema = (await SwaggerParser.bundle(input)) as OpenAPIObject;

  const ctx = generateContext(openApiSchema);

  const ast = new AstTsWriter();

  ast
    // import { initContract } from "@ts-rest/core";
    .add(tsNamedImport({ from: "@ts-rest/core", import_: ["initContract"] }))
    // import { z } from "zod";
    .add(tsNamedImport({ from: "zod", import_: ["z"] }))
    .add(tsNewLine())
    // const c = initContract();
    .add(tsVariableDeclaration("const", "c", { eq: tsFunctionCall("initContract") }))
    .add(tsNewLine());

  const schemasToExport = processSchemasForExport(ctx);

  // Generates the Zod schemas for each component schema.
  for (const { normalizedIdentifier, schema } of schemasToExport) {
    // const [identifier] = z.object({ ... }) | z.string() | z.number() | ...
    ast.add(
      tsVariableDeclaration("const", normalizedIdentifier, {
        eq: schemaObjectToAstZodSchema(schema, ctx),
      })
    );
  }

  ast.add(tsNewLine());

  if (schemasToExport.length > 0) {
    // export const schemas = { schema1, schema2, ... };
    ast
      .add(
        tsVariableDeclaration("const", "schemas", {
          eq: tsObject(
            ...schemasToExport.map(
              ({ normalizedIdentifier }) => [normalizedIdentifier] satisfies [string]
            )
          ),
          export_: true,
        })
      )
      .add(tsNewLine());
  }

  const apiOperationObjects = getApiOperationObjects(ctx);

  const tsRestAstContracts = apiOperationObjects.map((operationObject) =>
    apiOperationToAstTsRestContract(operationObject, ctx)
  );

  // export const contract = c.router({ ... });
  ast.add(
    tsVariableDeclaration("const", "contract", {
      eq: tsChainedMethodCall("c", ["router", tsObject(...tsRestAstContracts)]),
      export_: true,
    })
  );

  return await prettify(ast.toString(), prettierConfig);
}
