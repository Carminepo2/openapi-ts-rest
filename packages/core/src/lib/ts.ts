/**
 * This module provides a set of utility functions to generate TypeScript AST nodes.
 * It helps to avoid using the TypeScript compiler API directly, which can be verbose and error-prone.
 */

import { P, match } from "ts-pattern";
import {
  type ArrayLiteralExpression,
  type CallExpression,
  type ExportDeclaration,
  type Expression,
  type FalseLiteral,
  type Identifier,
  type ImportDeclaration,
  NodeFlags,
  type NullLiteral,
  type NumericLiteral,
  type ObjectLiteralExpression,
  type RegularExpressionLiteral,
  type StringLiteral,
  SyntaxKind,
  type TrueLiteral,
  type VariableStatement,
  factory,
  isExpression,
} from "typescript";

export type TsKeyword = "const" | "let" | "var";
export type TsLiteral = boolean | null | number | string;
export type TsLiteralOrExpression = Expression | TsLiteral;
export type TsFunctionCall = [identifier: string, ...args: TsLiteralOrExpression[]];

/**
 * Creates an AST TypeScript named import statement.
 *
 * @param import_ The names of the imports.
 * @param from The module specifier.
 * @returns The TypeScript AST import statement.
 *
 * @example
 * ```ts
 * tsNamedImport({ import_: ["foo", "bar"], from: "./module" });
 * // import { foo, bar } from "./module";
 * ```
 */
export function tsNamedImport({
  from,
  import_,
}: {
  from: string;
  import_: string[];
}): ImportDeclaration {
  return factory.createImportDeclaration(
    undefined,
    factory.createImportClause(
      false,
      undefined,
      factory.createNamedImports(
        import_.map((name) =>
          factory.createImportSpecifier(false, undefined, factory.createIdentifier(name))
        )
      )
    ),
    factory.createStringLiteral(from),
    undefined
  );
}

/**
 * Creates an AST TypeScript named export statement.
 *
 * @param export_ The identifiers to export
 * @returns The TypeScript AST named export statement.
 *
 * @example
 * ```ts
 * tsNamedExport({ export_: ["foo", "bar"] });
 * // export { foo, bar };
 * ```
 */
export function tsNamedExport({ export_: names }: { export_: string[] }): ExportDeclaration {
  return factory.createExportDeclaration(
    undefined,
    false,
    factory.createNamedExports(
      names.map((n) => factory.createExportSpecifier(false, undefined, factory.createIdentifier(n)))
    ),
    undefined,
    undefined
  );
}

/**
 * Accepts both Javascript primitive values and TypeScript AST nodes and returns the corresponding AST node.
 * If the input is a primitive value, it returns the corresponding TypeScript AST node.
 * If the input is already a TypeScript AST node, it returns the input as is.
 *
 * This function is useful when you want to accept both primitive values and TypeScript AST nodes as input.
 *
 * @param value - The primitive value or TypeScript AST node.
 * @returns The TypeScript AST node.
 *
 * @example
 * ```ts
 * tsLiteralOrExpression("foo");
 * // StringLiteral { text: "foo" }
 *
 * tsLiteralOrExpression(42);
 * // NumericLiteral { text: "42" }
 *
 * tsLiteralOrExpression(true);
 * // TrueLiteral {}
 *
 * tsLiteralOrExpression(StringLiteral { text: "foo" });
 * // StringLiteral { text: "foo" }
 * ```
 */
export function tsLiteralOrExpression(value: string): StringLiteral;
export function tsLiteralOrExpression(value: number): NumericLiteral;
export function tsLiteralOrExpression(value: true): TrueLiteral;
export function tsLiteralOrExpression(value: false): FalseLiteral;
export function tsLiteralOrExpression(value: null): NullLiteral;
export function tsLiteralOrExpression(value: Expression): Expression;
export function tsLiteralOrExpression(
  value: TsLiteralOrExpression
): Expression | FalseLiteral | NullLiteral | NumericLiteral | StringLiteral | TrueLiteral;
export function tsLiteralOrExpression(
  value: TsLiteralOrExpression
): Expression | FalseLiteral | NullLiteral | NumericLiteral | StringLiteral | TrueLiteral {
  return match(value)
    .with(P.string, (value) => factory.createStringLiteral(value))
    .with(
      P.number,
      (v) => v < 0,
      (v) =>
        factory.createPrefixUnaryExpression(
          SyntaxKind.MinusToken,
          factory.createNumericLiteral(Math.abs(v))
        )
    )
    .with(P.number, (value) => factory.createNumericLiteral(value))
    .with(true, () => factory.createTrue())
    .with(false, () => factory.createFalse())
    .with(null, () => factory.createNull())
    .when(isExpression, (v) => v)
    .exhaustive();
}

/**
 * Returns the corresponding TypeScript AST node for the given keyword.
 *
 * @param keyword - The keyword.
 * @returns The TypeScript AST node for the keyword.
 */
export function tsKeyword(keyword: "var"): NodeFlags.None;
export function tsKeyword(keyword: "let"): NodeFlags.Let;
export function tsKeyword(keyword: "const"): NodeFlags.Const;
export function tsKeyword(keyword: TsKeyword): NodeFlags;
export function tsKeyword(keyword: TsKeyword): NodeFlags {
  return match(keyword)
    .with("var", () => NodeFlags.None)
    .with("let", () => NodeFlags.Let)
    .with("const", () => NodeFlags.Const)
    .exhaustive();
}

/**
 * Creates an AST TypeScript variable declaration statement.
 *
 * @param keyword - The variable declaration keyword.
 * @param identifier - The variable identifier.
 * @param eq - The primitive value or the ast expresion to assign to the variable.
 * @param export_ - Whether to export the variable.
 *
 * @returns The TypeScript AST variable declaration statement.
 *
 * @example
 * ```ts
 * tsVariableDeclaration("const", "foo", { eq: "bar" });
 * // const foo = "bar";
 *
 * tsVariableDeclaration("let", "foo", { eq: tsObject({}) });
 * // let foo = {};
 *
 * tsVariableDeclaration("var", "foo", { eq: tsFunctionCall("fn", "arg"), _export: true });
 * // export var foo = fn("arg");
 * ```
 */
export function tsVariableDeclaration(
  keyword: TsKeyword,
  identifier: string,
  { eq: value, export_ }: { eq: TsLiteralOrExpression; export_?: boolean }
): VariableStatement {
  return factory.createVariableStatement(
    export_ ? [factory.createToken(SyntaxKind.ExportKeyword)] : undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier(identifier),
          undefined,
          undefined,
          tsLiteralOrExpression(value)
        ),
      ],
      tsKeyword(keyword)
    )
  );
}

/**
 * Creates an AST TypeScript function call expression.
 *
 * @param fn - The first argument is the identifier (name) of the function. The remaining arguments are the parameters to pass to the function.
 * @returns  An AST TypeScript function call expression.
 *
 * @example
 * ```ts
 * tsFunctionCall("fn", "arg1", "arg2");
 * // fn("arg1", "arg2");
 * ```
 */
export function tsFunctionCall(...fn: TsFunctionCall): CallExpression {
  const [identifier, ...fnCallArgs] = fn;
  return factory.createCallExpression(
    factory.createIdentifier(identifier),
    undefined,
    fnCallArgs.map(tsLiteralOrExpression)
  );
}

/**
 * Creates an AST TypeScript object literal expression.
 * @param properties An array of key-value pairs to create the object properties. The value can be primitive (which will be automatically converted to an AST expression) or an AST expression.
 * @returns The TypeScript AST object literal expression.
 *
 * @example
 * ```ts
 * tsObject(["foo", "bar"], ["baz", 42]);
 * // { foo: "bar", baz: 42 }
 * ```
 */
export function tsObject(
  ...properties: Array<[key: Identifier | string] | [key: string, value: TsLiteralOrExpression]>
): ObjectLiteralExpression {
  return factory.createObjectLiteralExpression(
    properties.map(([key, value]) => {
      if (value && typeof key === "string") {
        return factory.createPropertyAssignment(
          factory.createStringLiteral(key),
          tsLiteralOrExpression(value)
        );
      }
      return factory.createShorthandPropertyAssignment(
        typeof key === "string" ? factory.createIdentifier(key) : key,
        undefined
      );
    })
  );
}

/**
 * Creates an AST TypeScript array literal expression.
 * @param elements An array of elements to create the array. The elements can be primitive (which will be automatically converted to an AST expression) or an AST expression.
 * @returns The TypeScript AST array literal expression.
 *
 * @example
 * ```ts
 * tsArray("foo", 42);
 * // ["foo", 42]
 * ```
 */
export function tsArray(...elements: TsLiteralOrExpression[]): ArrayLiteralExpression {
  return factory.createArrayLiteralExpression(elements.map(tsLiteralOrExpression));
}

/**
 * Creates an AST TypeScript method call expression.
 * Method calls are chained together.
 *
 * @param identifier the identifier of the object to call the method on
 * @param param1 the method to call
 * @param chain the rest of the method calls
 * @returns the AST TypeScript property call expression
 *
 * @example
 * ```ts
 * tsPropertyCall("foo", ["bar", "baz"], ["qux", "quux"]);
 * // foo.bar("baz").qux("quux")
 * ```
 */
export function tsChainedMethodCall(identifier: string, ...chain: TsFunctionCall[]): Expression {
  if (identifier.length === 0) return factory.createIdentifier(identifier);

  return chain.reduce(
    (expression, [method, ...args]) =>
      factory.createCallExpression(
        factory.createPropertyAccessExpression(expression, factory.createIdentifier(method)),
        undefined,
        args.map(tsLiteralOrExpression)
      ),
    factory.createIdentifier(identifier) as Expression
  );
}

/**
 * Create an new line in the AST.
 * This might be a bit hacky, but it works.
 * @see https://stackoverflow.com/a/69240365
 *
 * @returns
 */
export function tsNewLine(): Identifier {
  return factory.createIdentifier("\n");
}

/**
 * Create an AST Typescript identifier.
 *
 * @param name the name of the identifier
 * @returns the AST Typescript identifier
 */
export function tsIdentifier(name: string): Identifier {
  return factory.createIdentifier(name);
}

/**
 * Create an AST Typescript regular expression literal.
 * @param pattern the pattern of the regular expression
 * @returns the AST Typescript regular expression literal
 */
export function tsRegex(pattern: string): RegularExpressionLiteral {
  return factory.createRegularExpressionLiteral(pattern);
}
