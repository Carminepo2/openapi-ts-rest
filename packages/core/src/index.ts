import { apiOperationToAstTsRestContract } from "./converters/apiOperationToAstTsRestContract.js";
import { generateContext } from "./context.js";
import { getAPIOperationsObjects } from "./getAPIOperationsObjects.js";
import { prettify } from "./lib/prettier.js";
import {
  tsVariableDeclaration,
  tsFunctionCall,
  tsNamedImport,
  tsNewLine,
  tsObject,
  tsChainedMethodCall,
} from "./lib/ts.js";
import { AstTsWriter } from "./lib/utils.js";
import { schemaObjectToAstZodSchema } from "./converters/schemaObjectToAstZodSchema.js";
import SwaggerParser from "@apidevtools/swagger-parser";
import type { OpenAPIObject } from "openapi3-ts/oas30";

interface GenerateTsRestContractFromOpenAPIOptions {
  input: string;
}

/**
 * Generates a ts-rest contract from an OpenAPI schema.
 * @param {GenerateTsRestContractFromOpenAPIOptions} options - The options for the generation.
 */
export async function generateTsRestContractFromOpenAPI({
  input,
}: GenerateTsRestContractFromOpenAPIOptions): Promise<string> {
  const openApiSchema = (await SwaggerParser.bundle(input)) as OpenAPIObject;

  const ctx = generateContext(openApiSchema);

  const ast = new AstTsWriter();

  ast
    // import { initContract } from "@ts-rest/core";
    .add(tsNamedImport({ import_: ["initContract"], from: "@ts-rest/core" }))
    // import { z } from "zod";
    .add(tsNamedImport({ import_: ["z"], from: "zod" }))
    .add(tsNewLine())
    // const c = initContract();
    .add(tsVariableDeclaration("const", "c", { eq: tsFunctionCall("initContract") }))
    .add(tsNewLine());

  // Generates the Zod schemas for each component schema.
  for (const { normalizedIdentifier, schema } of ctx.topologicallySortedSchemas) {
    // const [identifier] = z.object({ ... }) | z.string() | z.number() | ...
    ast.add(tsVariableDeclaration("const", normalizedIdentifier, { eq: schemaObjectToAstZodSchema(schema, ctx) }));
  }

  ast.add(tsNewLine());

  const schemaIdentifiersToExport = ctx.topologicallySortedSchemas
    .filter(({ ref }) => ctx.schemasToExportMap.has(ref))
    .map(({ normalizedIdentifier }) => [normalizedIdentifier] satisfies [string]);

  if (schemaIdentifiersToExport.length > 0) {
    // export const schemas = { schema1, schema2, ... };
    ast
      .add(tsVariableDeclaration("const", "schemas", { eq: tsObject(...schemaIdentifiersToExport), export_: true }))
      .add(tsNewLine());
  }

  // Gets the API operations objects from the OpenAPI schema, which are used to generate each contract.
  const operationObjects = getAPIOperationsObjects(ctx);

  const tsRestAstContracts = operationObjects.map((operationObject) => {
    return apiOperationToAstTsRestContract(operationObject, ctx);
  });

  // export const contract = c.router({ ... });
  ast.add(
    tsVariableDeclaration("const", "contract", {
      eq: tsChainedMethodCall("c", ["router", tsObject(...tsRestAstContracts)]),
      export_: true,
    })
  );

  return await prettify(ast.toString());
}
