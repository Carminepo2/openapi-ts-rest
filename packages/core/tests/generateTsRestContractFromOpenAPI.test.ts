import ts from "typescript";

import { generateTsRestContractFromOpenAPI } from "../src/generateTsRestContractFromOpenAPI";

const FIXTURES = [
  "https://raw.githubusercontent.com/OAI/OpenAPI-Specification/main/examples/v3.0/petstore.yaml",
  "https://raw.githubusercontent.com/teamdigitale/api-openapi-samples/master/openapi-v3/defibrillatori-example.yaml",
];

describe("generateTsRestContractFromOpenAPI", () => {
  it.each(FIXTURES)("should successfully transpile module without ts errors", async (input) => {
    const module = await generateTsRestContractFromOpenAPI({
      input,
    });

    const output = ts.transpileModule(module, {
      compilerOptions: {
        module: ts.ModuleKind.NodeNext,
        noEmit: true,
        strict: true,
        target: ts.ScriptTarget.ESNext,
      },
      reportDiagnostics: true,
    });

    expect(output.diagnostics).toHaveLength(0);
  });

  it.each(FIXTURES)("should match snapshot", async (input) => {
    const module = await generateTsRestContractFromOpenAPI({
      input,
    });

    expect(module).toMatchSnapshot();
  });
});
