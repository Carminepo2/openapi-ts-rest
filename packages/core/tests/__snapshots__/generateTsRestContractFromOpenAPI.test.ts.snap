// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateTsRestContractFromOpenAPI > should match snapshot for https://raw.githubusercontent.com/OAI/OpenAPI-Specification/main/examples/v3.0/petstore.yaml 1`] = `
"import { initContract } from "@ts-rest/core";
import { z } from "zod";

const c = initContract();

const Pet = z.object({
  id: z.number().int(),
  name: z.string(),
  tag: z.string().optional(),
});
const Pets = z.array(Pet).max(100);
const Error = z.object({ code: z.number().int(), message: z.string() });

export const schemas = { Pet, Pets, Error };

export const contract = c.router({
  listPets: {
    method: "GET",
    path: "/pets",
    summary: "List all pets",
    query: z.object({ limit: z.number().int().lte(100).optional() }),
    responses: {
      "200": Pets,
      "400": Error,
      "401": Error,
      "403": Error,
      "404": Error,
      "405": Error,
      "409": Error,
      "415": Error,
      "500": Error,
    },
  },
  createPets: {
    method: "POST",
    path: "/pets",
    summary: "Create a pet",
    body: Pet,
    contentType: "application/json",
    responses: {
      "201": z.void(),
      "400": Error,
      "401": Error,
      "403": Error,
      "404": Error,
      "405": Error,
      "409": Error,
      "415": Error,
      "500": Error,
    },
  },
  showPetById: {
    method: "GET",
    path: "/pets/:petId",
    summary: "Info for a specific pet",
    pathParams: z.object({ petId: z.string() }),
    responses: {
      "200": Pet,
      "400": Error,
      "401": Error,
      "403": Error,
      "404": Error,
      "405": Error,
      "409": Error,
      "415": Error,
      "500": Error,
    },
  },
});
"
`;

exports[`generateTsRestContractFromOpenAPI > should match snapshot for https://raw.githubusercontent.com/teamdigitale/api-openapi-samples/master/openapi-v3/defibrillatori-example.yaml 1`] = `
"import { initContract } from "@ts-rest/core";
import { z } from "zod";

const c = initContract();

const Problem = z.object({
  detail: z
    .string()
    .max(4096)
    .regex(/^.{0,1024}$/)
    .optional(),
  instance: z.string().max(2048).url().optional(),
  status: z.number().int().gte(100).lt(600).optional(),
  title: z
    .string()
    .max(64)
    .regex(/^[ -~]{0,64}$/)
    .optional(),
  type: z.string().max(2048).url().optional().default("about:blank"),
});
const DefibrillatoreId = z
  .string()
  .min(36)
  .max(36)
  .regex(/[a-z0-9-]{36}/)
  .uuid();
const Defibrillatore = z.object({
  id: DefibrillatoreId,
  status: z.enum(["attivo", "non-attivo"]),
  gestore: z
    .string()
    .max(32)
    .regex(/[a-zA-Z0-9\\-_]+/),
});
const GeoJSONFeature = z.object({
  type: z.literal("Feature"),
  geometry: z.object({
    type: z.literal("Point"),
    coordinates: z.array(z.number().gte(-180).lte(180)).min(2).max(2),
  }),
});
const DefibrillatoreGeo = z.object({
  type: z.literal("Feature"),
  geometry: z.object({
    type: z.literal("Point"),
    coordinates: z.array(z.number().gte(-180).lte(180)).min(2).max(2),
  }),
  properties: Defibrillatore,
});
const DefibrillatoriGeo = z.object({
  type: z.literal("FeatureCollection"),
  features: z.array(DefibrillatoreGeo).min(0).max(1000),
});
const AreaAmministrativa = z
  .string()
  .min(2)
  .max(6)
  .regex(/[0-9]{2,6}/);
const Latitude = z.number().gte(-90).lte(90);
const Longitude = z.number().gte(-180).lte(180);

export const schemas = {
  Problem,
  DefibrillatoreId,
  Defibrillatore,
  GeoJSONFeature,
  DefibrillatoreGeo,
  DefibrillatoriGeo,
  AreaAmministrativa,
  Latitude,
  Longitude,
};

export const contract = c.router({
  mostraStatus: {
    method: "GET",
    path: "/status",
    summary: "Get status",
    responses: {
      "200": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "400": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "401": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "403": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "404": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "405": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "409": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "415": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "500": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
    },
  },
  listaDefibrillatori: {
    method: "GET",
    path: "/defibrillatori",
    summary: "Elenca defibrillatori",
    query: z.object({
      codice_istat: z
        .string()
        .min(2)
        .max(6)
        .regex(/[0-9]{2,6}/)
        .optional(),
      lat: z.number().gte(-90).lte(90).optional(),
      lon: z.number().gte(-180).lte(180).optional(),
      radius: z.number().int().gte(0).lte(2000).optional().default(500),
    }),
    responses: {
      "200": c.otherResponse({
        contentType: "application/geo+json",
        body: DefibrillatoriGeo,
      }),
      "400": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "401": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "403": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "404": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "405": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "409": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "415": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "500": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
    },
  },
  creaDefibrillatore: {
    method: "POST",
    path: "/defibrillatori",
    summary: "Aggiungi defibrillatore",
    body: DefibrillatoreGeo,
    contentType: "application/json",
    responses: {
      "201": c.otherResponse({
        contentType: "application/geo+json",
        body: DefibrillatoreGeo,
      }),
      "400": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "401": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "403": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "404": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "405": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "409": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "415": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "500": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
    },
  },
  mostraDefibrillatore: {
    method: "GET",
    path: "/defibrillatori/:id",
    summary: "Elenca un defibrillatore",
    pathParams: z.object({
      id: z
        .string()
        .min(36)
        .max(36)
        .regex(/[a-z0-9-]{36}/)
        .uuid(),
    }),
    responses: {
      "200": DefibrillatoreGeo,
      "400": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "401": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "403": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "404": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "405": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "409": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "415": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "500": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
    },
  },
  modificaDefibrillatore: {
    method: "PATCH",
    path: "/defibrillatori/:id",
    summary: "Aggiorna un defibrillatore",
    pathParams: z.object({
      id: z
        .string()
        .min(36)
        .max(36)
        .regex(/[a-z0-9-]{36}/)
        .uuid(),
    }),
    body: DefibrillatoreGeo,
    contentType: "application/json",
    responses: {
      "200": c.otherResponse({
        contentType: "application/geo+json",
        body: DefibrillatoreGeo,
      }),
      "403": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "400": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "401": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "404": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "405": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "409": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "415": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "500": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
    },
  },
  eliminaDefibrillatore: {
    method: "DELETE",
    path: "/defibrillatori/:id",
    summary: "Elimina un defibrillatore.",
    pathParams: z.object({
      id: z
        .string()
        .min(36)
        .max(36)
        .regex(/[a-z0-9-]{36}/)
        .uuid(),
    }),
    body: z.void(),
    responses: {
      "204": z.void(),
      "400": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "401": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "403": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "404": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "405": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "409": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "415": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
      "500": c.otherResponse({
        contentType: "application/problem+json",
        body: Problem,
      }),
    },
  },
});
"
`;
